### DISCUSSION TITLE
DataGrid component

### DISCUSSION THREAD

**Discussion by dukesx** [mantinedev/mantine#1057] [Feature requests]
I was scrolling through, just to encounter a few libraries i needed to use for an admin panel i am developing currently. i needed data tables to show data.  so i came across these two: 

1. Revo Grid
2. React Data Grid

Two of the most lightest and actually "data" centered tables that i have seen by far. Now i am not sure if they go with CSS in JS or emotion but worth a try if anyone from the community is willing to work on it. Its probably a needful implementation considering that data tables are very commonly used. Both libraries have been updated constantly, according to OpenBase Graph, active team. 

- Screen 1 : Revo Grid
![ss1](https://user-images.githubusercontent.com/5988991/135898980-3e871cf2-a7b7-44d6-b11c-b78ea3437ecc.png)

- Screen 2: React Data Grid
![ss2](https://user-images.githubusercontent.com/5988991/135898989-e8f87e75-a7d1-4114-a15f-b491ff41e6ec.png)

**Comment by dukesx** [comment]
> I'll be working on data grid package some time next year, most likely it will be built with react-table – https://react-table.tanstack.com/

Well, isn't this library going to take an immense amount of time ?  i like the fact that it's headless and has hooks but implementation seems complex somewhat.

**Comment by rtivital** [comment]
> Well, isn't this library going to take an immense amount of time

Why? react-table handles all complex cases. Headless approach will allow us to customize styles to match other components while handling all crazy stuff like columns resizing.

**Comment by esistgut** [comment]
Any update on this? The lack of a data table component is a massive blocker for any kind of CRUD application.

**Comment by wapenshaw** [comment]
I'd like to add a suggestion regarding datatables and datagrids, please make them responsive, i.e. the columns stack into rows on mobile screen sizes. [Example](https://css-tricks.com/responsive-data-tables/)

There are very few libraries in the react ecosystem that offer this functionality (afaik, primereact, elasticui) Last time I checked even Ant Design didn't have responsive tables (they can hide/show columns based on breakpoints but thats it) 

If you look at ElasticUI, all their tables and grid components offer this feature but EUI is not nearly as flexible as mantine 

https://elastic.github.io/eui/#/tabular-content/tables#a-basic-table << check this page on a mobile screen width to see the stacking

**Reply by mlarcher** [reply]
I agree that this feature is a really nice one, and should be a target for Mantine to keep a great mobile experience !

**Comment by alexkreidler** [comment]
Personally I really love this project: https://github.com/glideapps/glide-data-grid, which uses React but renders using Canvas instead of regular HTML, which may limit some the styling customization options.

**Comment by AnhV90** [comment]
Any update on this? I'm about to implement react-table myself.

**Reply by mvmilanjain** [reply]
Please refer https://codesandbox.io/s/react-table-datagrid-eojw8

**Reply by dukesx** [reply]
You Sir are the REAL MVP. Bravo !! Good Job 👍 👍 🎉 Can't wait to see the PR.

**Reply by Kuechlin** [reply]
i'm also implementing a data grid with react-table v8 (it now has good typescript support).

https://github.com/Kuechlin/mantine-data-grid

**Reply by dukesx** [reply]
> i'm also implementing a data grid with react-table v8 (it now has good typescript support).
> 
> https://github.com/Kuechlin/mantine-data-grid

Nice ! looks beautiful !

**Reply by moshOntong-IT** [reply]
> i'm also implementing a data grid with react-table v8 (it now has good typescript support).
> 
> https://github.com/Kuechlin/mantine-data-grid

You are savior woaahhh

**Reply by rwieruch** [reply]
Another alternative: Check out the [Mantine integration](https://react-table-library.com/?path=/docs/library-themes-mantine--base) in react-table-library.

**Reply by codellyson** [reply]
> Mantine integration
Thanks you literally saved me tons of work

**Comment by mlarcher** [comment]
@rtivital any idea when you'll have time to work on that ? Is it something to be expected soon, in v5, not before next year or more...? Any hint would be appreciated

**Reply by rtivital** [reply]
I'm not planning to work on DataGrid any time soon

**Reply by mlarcher** [reply]
Thanks for the quick reply 👍🏻

**Reply by razfriman** [reply]
Any chance you will reconsider by adapting the above 2 solutions as an official datagrid component directly into Mantine?

**Reply by rtivital** [reply]
I'm not planning to work on DataGrid any time soon

**Reply by elarens** [reply]
@rtivital Is there anything that can be done by the community to accelerate the development of this component?

**Reply by iamyuu** [reply]
https://github.com/Kuechlin/mantine-data-grid

**Reply by elarens** [reply]
@iamyuu I've seen the impressive effort by other people to build the component but I would love to see it an official part of the library, hence my question to Vitaly

**Reply by rtivital** [reply]
No, there are no plans to add such component in the next six months

**Comment by icflorescu** [comment]
I know there are already a few alternatives out there, but you might also want to take a look at [Mantine DataTable](https://icflorescu.github.io/mantine-datatable/), a component I've built with Mantine primitives (not React Table), that covers most of the features mentioned above and should support:

- Asynchronous data loading
- Pagination
- Multiple rows selection
- Scrollable behavior, while maintaining a fixed header, pagination footer and row selection column
- Column sorting
- Custom cell data rendering
- Row context-menu
- Dark themes

I've also tried to build a nice documentation website (following as much as possible the excellent Mantine's docs design patterns) and I've included a series of examples and usage scenarios.
The component properties are be self explanatory, well typed and also documented with JSDoc, so it should be quite straightforward to use.

[![Mantine DataTable](https://user-images.githubusercontent.com/581999/187723353-0e082dbc-925b-4d0d-a14e-94c00073ee08.png)](https://icflorescu.github.io/mantine-datatable/)

**Reply by dukesx** [reply]
@icflorescu Thank you ! @rtivital , if you can't do it, i don't see why you shouldn't incorporate "THIS" table in your components library, considering how many people have requested this. So far, so good. It's as mantine as it gets. When the time comes, you can either create from scratch or update this table with the react-table hooks library you were talking about.

**Reply by rtivital** [reply]
I do not want to do it now because it will require a lot of effort to maintain, use third-party lib for now

**Reply by icflorescu** [reply]
@dukesx - thank you for your kind words!
@rtivital - totally agree
Anyway, `Mantine DataTable` is still a bit rough around the edges, but I'll do my best to improve it, maintain it and keep it *as mantine as it gets*.

**Reply by icflorescu** [reply]
...and, by the way, any help (i.e. starring the repo 🙏, raising issues and **especially** coming up with meaningful PRs ❤️) would be more than welcome :-)

**Reply by dukesx** [reply]
> @dukesx - thank you for your kind words!
> @rtivital - totally agree
> Anyway, `Mantine DataTable` is still a bit rough around the edges, but I'll do my best to improve it, maintain it and keep it *as mantine as it gets*.

Looking forward mate. It's fantastic. Truly phenomenal work. It has everything a "normal" yet complex use case could ever need.

**Reply by Bladeyy** [reply]
@icflorescu Damn truly a great work you did there. I was casually checking if there's any update on mantine data grid work. Glad I checked else I would've missed this. Like @dukesx said it has ticked all normal yet complex use cases. Truly a great work. Hope this gets better by time with more features and later incorporated to mantine. Thanks dude.

**Reply by Kuechlin** [reply]
Hey,

Why haven't you used react-table and instead build it yourself with lodash?

With react table you only have to build the UI like I did with [mantine-data-grid](https://kuechlin.github.io/mantine-data-grid/)

**Reply by icflorescu** [reply]
@Kuechlin - I built it "from scratch" because I wanted it to be "_as mantine as it gets_", esthetically speaking, and cover the basic use-cases without having to fight my way around another library. I also wanted to maintain a low bundle size. I'll also get rid of `lodash` peer dependency in a day or two :-).

**Reply by icflorescu** [reply]
As of [Mantine-DataTable](https://icflorescu.github.io/mantine-datatable/) `v1.1.0`, lodash peer dependency is no longer necessary 🎉

**Reply by retromack** [reply]
Great work @icflorescu

**Reply by icflorescu** [reply]
Starting with v.1.7.0, `Mantine DataTable` features [row expansion](https://icflorescu.github.io/mantine-datatable/examples/expanding-rows).

**Reply by elarens** [reply]
@icflorescu Marvellous implementation of a very useful feature and excellent documentation!
I can see that the rows are clickable (which makes sense) - what happens if there are other clickable components in the row, like links or action buttons? How will they interact?

**Reply by icflorescu** [reply]
Thanks for the kind words, @elarens!

Rows are indeed clickable if you are:
- [providing an `onRowClick` handler](https://icflorescu.github.io/mantine-datatable/examples/handling-row-clicks);
- using a [context menu triggered by left clicking](https://icflorescu.github.io/mantine-datatable/examples/row-context-menu);
- using the [row expansion](https://icflorescu.github.io/mantine-datatable/examples/expanding-rows) feature.

In my opinion, you simply shouldn't place clickable components in a row if you're using one of the above features because you'd end up confusing the user. If, for whatever reason, you absolutely **have to**, you could intercept the `click` event on the clickable component and do a `.stopPropagation()` on it. To my knowledge this works even for simple links.

Let's [discuss here](https://github.com/icflorescu/mantine-datatable/discussions) if you feel like this should somehow be explained on the docs website in an example.

**Reply by elarens** [reply]
Posted a question in discussion #28

**Reply by icflorescu** [reply]
@elarens, please have a look at **Using with clickable components in cells** - currently the last section on [this example page](https://icflorescu.github.io/mantine-datatable/examples/expanding-rows).

**Reply by icflorescu** [reply]
One major key thing I'd like to add about [Mantine DataTable](https://icflorescu.github.io/mantine-datatable), now that more than 6 months have passed since its initial release: despite its large number of features and out-of-the-box Mantine look-and-feel, it has no external dependencies and it's probably the lightest alternative in terms of bundle size:

[![Mantine DataTable brings datagrid-like functionality to your data-rich user interfaces](https://user-images.githubusercontent.com/581999/230039639-a1ba8a4d-fbe5-4447-bf39-24f7b7e7d8c5.png)](https://icflorescu.github.io/mantine-datatable)

Though it doesn't have the full feature set of React Table, it's built solely with Mantine primitives and should cover most usage scenarios. It's backed by a nice custom-built mobile-friendly [documentation website](https://icflorescu.github.io/mantine-datatable).

**Comment by icflorescu** [comment]
@Bladeyy
> Hope this gets better by time with more features

That's the plan.

**Comment by rwieruch** [comment]
I created [React Table Library](https://react-table-library.com/) as an almost headless table library which easily integrates with UI libraries such as Mantine. In the documentation, you will find Mantine via [Library Themes/Mantine](https://react-table-library.com/?path=/docs/library-themes-mantine--base).

It comes with:

- TypeScript support
- Tailwind CSS support
- custom UI library support (Mantine, MUI, ...)
- server-side operations support (e.g. server-side pagination)
- virtualization
- fixed columns/header
- sorting/filtering/searching
- tree view
- single-select/multi-select
- pagination
- custom theming
- column hiding
- ... and much more ...

More [awesome table examples](https://react-tables.com/) built with the library.

![Screenshot 2022-09-17 at 12 57 52](https://user-images.githubusercontent.com/2479967/190853296-b7c17d8c-3503-45e4-a2a8-528f139ad03c.jpg)

**Comment by vaynevayne** [comment]
https://www.mantine-react-table.com/docs/examples/data-export#data-export-example , i want try this

**Comment by ChinmayGopal931** [comment]
@icflorescu is there anyway to add styling to each row in a dataGrid? I want each row to be styled differently

**Reply by icflorescu** [reply]
@ChinmayGopal931: Yes, individual row styling is possible.

Have a look at [the last section ("Row styling") in the "Additional styling" example](https://icflorescu.github.io/mantine-datatable/examples/additional-styling#row-styling) of the documentation website.

[The documentation website](https://icflorescu.github.io/mantine-datatable) includes lots of examples of what you can do with Mantine DataTable:

[![Mantine DataTable brings datagrid-like functionality to your data-rich user interfaces](https://user-images.githubusercontent.com/581999/230036979-0283a71c-abd7-4840-a398-411a53d0bdfd.png)](https://icflorescu.github.io/mantine-datatable)
